### ISSUE TITLE
Disable timeout (not zero, not infinite, but disable it)

### ISSUE THREAD

**Issue by SengerM** [pyvisa/pyvisa#322]
Hello,

I would like to know if it is possible to use `instrument.write(...)` and continue the execution of the program without waiting for any kind of response from the instrument. 

I need this because I am using two voltmeters that need to be driven from the same trigger signal. Thus I 
1) Disable trigger from the source.
2) Set trigger arm in voltmeter one.
3) Set trigger arm in voltmeter two.
4) Activate trigger signal from the source.
Under this conditions both voltmeters should trigger at the same time. But it happens that in step 2 I am getting a timeout and the program stops.

The relevant part of the code is as follows:

HP33120A.write('OUTP:SYNC OFF') # Disable the sync terminal from the function generator
HP3458A_1.write('TARM SGL') # Enable sampling, single shot when trigger.
HP3458A_2.write('TARM SGL') # Enable sampling, single shot when trigger.
HP33120A.write('OUTP:SYNC ON') # Enable the sync terminal so the voltmeters trigger.

and the program is ending in the second line due to the timeout.

Thanks in advance,
Matías

**Comment by MatthieuDartiailh** [comment]
I have several questions. 
First can you actually communicate at all with your instrument ? For example does your instrument answers to `instr.query('ID?')`  (as it should per the documentation https://www.keysight.com/upload/cmc_upload/All/EPSG084364.pdf) ? If it does not first we need to fix this, and the answer may be to properly set the termination characters used for reading and writing. Also could you check your command syntax ? According to the documentation the TARM command may expect number_arms following SGL.
Second could you post the output of `python -m visa info`, so that I know on which OS and with which version of pyvisa you are working.
Third could you post your full script ? It may help debugging this issue.

Basically I believe you are mostly seeing the result of a real issue in communicating with the device and that even if pyvisa was not emitting a timeout, your instrument would not behave as you expect.

**Comment by SengerM** [comment]
Thanks for the quick answer. Here is the code I am using (cannot find how to attach it as a file):

import visa
import numpy as np
import math
import matplotlib.pyplot as plt
plt.rc('axes', axisbelow=True) # This is for plotting data in front of the grid.
plt.rcParams.update({'font.size': 8}) # Changes the default font size for the plots.
from time import sleep
import struct # This is used in parsing bynary data.
import scipy.optimize as opt # This is used for fiting models and data.
import pymagpy as mg # This is a custom class I have created.
from utils.fit_bootstrap import *
from utils.HP3458A_lib import *
from utils.timestamp import *

def HP3458A_configure_DCV_digitalizing(HP3458A, samples_per_burst, effective_sampling_frequency):
	HP3458A.write('PRESET DIG')
	HP3458A.write('MEM FIFO') # ENABLE READING MEMORY, FIFO MODE.
	HP3458A.write('MFORMAT SINT') # SINT READING MEMORY FORMAT.
	HP3458A.write('OFORMAT SINT') # Set the output format when reading the samples in memory.
	HP3458A.write('TIMER ' + str(1/effective_sampling_frequency)) # Specifies sampling frequency.
	HP3458A.write('APER 500e-9') # Set aper time to 500 ns which gives a resolution of 16 bits (see http://literature.cdn.keysight.com/litweb/pdf/03458-90014.pdf table 5-2).
	HP3458A.write('NRDGS ' + str(samples_per_burst) + ', 6') # Specifies the number of samples to be recorded and the event in which to sample (6 is for "timer").
	HP3458A.write('SSRC EXT') # Set external trigger.


# REFERENCES: 
# 	[1] Keysight 3458A Multimeter, https://xdevs.com/doc/HP_Agilent_Keysight/3458A/03458-90014.pdf

# PROGRAM PARAMETERS ----------------------
GENERATOR_FREQUENCY = 1e3 # Hertz
GENERATOR_AMPLITUDE = 1 # Volt
SAMPLES_PER_BURST = 10000 # Integer between 1 and ?? (have not found yet the size of the memory.
CYCLES_PER_BURST = 100 # Number of signal cycles to be recorded in each burst.
EFFECTIVE_SAMPLING_FREQUENCY = SAMPLES_PER_BURST*GENERATOR_FREQUENCY/CYCLES_PER_BURST
SAVE_PLOTS = 0
# -----------------------------------------

# print('\t|\t|\t|\t|\t|\n'*10)

timestamp = get_timestamp()
print('Timestamp: ' + timestamp)

set_frequency = mg.magnitude(GENERATOR_FREQUENCY, GENERATOR_FREQUENCY*20e-6) # The value for the uncertainty in the generator frequency was taken from https://www.keysight.com/upload/cmc_upload/All/6C0633120A_USERSGUIDE_ENGLISH.pdf
set_amplitude = mg.magnitude(GENERATOR_AMPLITUDE, GENERATOR_AMPLITUDE*1e-2) # The value for the uncertainty in the generator amplitude was taken from https://www.keysight.com/upload/cmc_upload/All/6C0633120A_USERSGUIDE_ENGLISH.pdf

rm = visa.ResourceManager()

HP3458A_1 = rm.open_resource('GPIB0::22::INSTR') # Multimeter.
HP3458A_1.read_termination = '\r' # Each reading output to the GPIB in ASCII format is normally followed by '\r\n' (see [1], page 99).
HP3458A_2 = rm.open_resource('GPIB0::23::INSTR') # Multimeter.
HP3458A_2.read_termination = '\r' # Each reading output to the GPIB in ASCII format is normally followed by '\r\n' (see [1], page 99).
HP33120A = rm.open_resource('GPIB0::10::INSTR') # Function generator.

# Check that the instruments are responding:
print('HP3458A_1.query(\'ID?\') returns --> ' + HP3458A_1.query('ID?'))
print('HP3458A_2.query(\'ID?\') returns --> ' + HP3458A_2.query('ID?'))
print('HP33120A.query(\'*IDN?\') returns --> ' + HP33120A.query('*IDN?'))
# Reset the instruments:
HP3458A_1.write('RESET')
HP3458A_2.write('RESET')
HP33120A.write('*RST')
print('Setting generator output to ' + str(set_amplitude) + ' V and frequency to ' + str(set_frequency) + ' Hz')
HP33120A.write('APPL:SIN {}, {}, 0'.format(GENERATOR_FREQUENCY, GENERATOR_AMPLITUDE)) # Set a sine wave in the generator.
# HP33120A.write('OUTP:SYNC OFF') # Disable the sync terminal (so the voltmeters do not trigger until they are configured).
print('Configuring voltmeters...')
HP3458A_configure_DCV_digitalizing(HP3458A_1, SAMPLES_PER_BURST, EFFECTIVE_SAMPLING_FREQUENCY)
HP3458A_configure_DCV_digitalizing(HP3458A_2, SAMPLES_PER_BURST, EFFECTIVE_SAMPLING_FREQUENCY)
HP3458A_1.write('TARM SGL') # ENABLE SAMPLING (SGL specifies to trigger a single shot, see [1] pag. 251).
HP3458A_2.write('TARM SGL') # ENABLE SAMPLING (SGL specifies to trigger a single shot, see [1] pag. 251).
print('Waiting...')
sleep(1)
print('Enabling generator\'s trigger output...')
HP33120A.write('OUTP:SYNC ON') # Enable the sync terminal so the voltmeters trigger.
print('Waiting...')
sleep(1)
print('Reading data from voltmeter 1...')
HP3458A_1.write('RMEM 1,{}'.format(SAMPLES_PER_BURST))
samples_1 = np.asarray(struct.unpack('>'+'h'*SAMPLES_PER_BURST, HP3458A_1.read_bytes(2*SAMPLES_PER_BURST))) # '>h' denotes "16 bit integer in two's complement and big endian"
samples_1 = samples_1*float(HP3458A_1.query('ISCALE?')) # Conversion factor from SINT format to voltage values.
# samples_1 = np.flipud(samples_1) # Samples come reversed in time.
samples_1 = samples_1*-1
print('Reading data from voltmeter 2...')
HP3458A_2.write('RMEM 1,{}'.format(SAMPLES_PER_BURST))
samples_2 = np.asarray(struct.unpack('>'+'h'*SAMPLES_PER_BURST, HP3458A_2.read_bytes(2*SAMPLES_PER_BURST))) # '>h' denotes "16 bit integer in two's complement and big endian"
samples_2 = samples_2*float(HP3458A_2.query('ISCALE?')) # Conversion factor from SINT format to voltage values.
# samples_2 = np.flipud(samples_2) # Samples come reversed in time.
samples_2 = samples_2*-1
# PLOT -----------------------------------------------
print('Plotting sampled data...')
FIG_WIDTH = 130e-3 # Figure width in meters (size of the chart+axis_ticks, not counting legend, title, etc.).
FIG_RATIO = [1, 0.8] # XY ratio aspect of figures.
f, (ax1, ax2) = plt.subplots(2, sharex=True, figsize=(FIG_WIDTH*FIG_RATIO[0]/25.4e-3, FIG_WIDTH*FIG_RATIO[1]/25.4e-3)) # Create the figure for plotting.
f.subplots_adjust(hspace=0.3) # Fine-tune figure; make subplots close to each other and hide x ticks for all but bottom plot.
ax1.plot(samples_1, marker='.', label='V1')
ax1.plot(samples_2, marker='.', label='V2')
ax1.set_xlabel('Sample number')
ax1.set_ylabel('Voltage (V)')
ax1.grid(b=True, which='major', color='#aaaaaa', linestyle='-', linewidth=0.5) # Configure the grid in a more fachera way.
ax1.grid(b=True, which='minor', color='#dddddd', linestyle='-', linewidth=0.25) # Configure the grid in a more fachera way.
ax1.minorticks_on() # Enables minor ticks without text, only the ticks.
ax1.legend()
ax2.plot(samples_1-samples_2, marker='.', label='V1-V2')
ax2.set_xlabel('Sample number')
ax2.set_ylabel('Voltage (V)')
ax2.grid(b=True, which='major', color='#aaaaaa', linestyle='-', linewidth=0.5) # Configure the grid in a more fachera way.
ax2.grid(b=True, which='minor', color='#dddddd', linestyle='-', linewidth=0.25) # Configure the grid in a more fachera way.
ax2.minorticks_on() # Enables minor ticks without text, only the ticks.
ax2.legend()
plt.draw()
if SAVE_PLOTS:
	plt.savefig(timestamp+'_pure_signals.pdf', bbox_inches='tight')
# ----------------------------------------------------

# ESTIMATION OF fs1 AND fs2 -----------------------------------------------------------
y_data = samples_1 - samples_2 # 'y' data to fit.
x_data = np.arange(0, len(y_data), 1) # 'x' data to fit.
def fit_func(n, A, B): # 'A and B are the parameters, 'n' is the 'x' axis.
    return 2*B*n*np.cos(A*n)
def fit_func_ff(x, p):
    return fit_func(x, *p)
# Estimation of the frequency using FFT ------------------------------------------------
fft_diff = np.fft.rfft(y_data) # Compute the FFT.
A0 = np.linspace(0, np.pi, len(fft_diff))[np.argmax(np.absolute(fft_diff))] # Estimation for the parameter 'A' in the initial guess when aplying least squares.
# Estimation of fs1 and fs2 by leasqr, taken from here: https://stackoverflow.com/questions/14581358/getting-standard-errors-on-fitted-parameters-using-the-optimize-leastsq-method-i
p0 = [A0, np.amax(y_data)/len(y_data)] # Initial params.
pfit, perr = fit_bootstrap(p0, x_data, y_data, fit_func_ff)
A = mg.magnitude(pfit[0], perr[0])
B = mg.magnitude(pfit[1], perr[1])
fs1 = 2*np.pi*set_frequency/(A+B) 
fs2 = 2*np.pi*set_frequency/(A-B)
print('fs1 = ' + str(fs1))
print('fs2 = ' + str(fs2))

# PLOT FITTED FUNCTION -----------------------------------------------
FIG_WIDTH = 130e-3 # Figure width in meters (size of the chart+axis_ticks, not counting legend, title, etc.).
FIG_RATIO = [1, 0.5] # XY ratio aspect of figures.
f, (ax1) = plt.subplots(1, sharex=True, figsize=(FIG_WIDTH*FIG_RATIO[0]/25.4e-3, FIG_WIDTH*FIG_RATIO[1]/25.4e-3)) # Create the figure for plotting.
f.subplots_adjust(hspace=0.3) # Fine-tune figure; make subplots close to each other and hide x ticks for all but bottom plot.
ax1.plot(samples_1-samples_2, marker='.', label='Data')
ax1.plot(fit_func(x_data, A.val, B.val), label='Fit')
ax1.set_xlabel('Sample number')
ax1.set_ylabel('Voltage (V)')
ax1.grid(b=True, which='major', color='#aaaaaa', linestyle='-', linewidth=0.5) # Configure the grid in a more fachera way.
ax1.grid(b=True, which='minor', color='#dddddd', linestyle='-', linewidth=0.25) # Configure the grid in a more fachera way.
ax1.minorticks_on() # Enables minor ticks without text, only the ticks.
ax1.legend()
plt.draw()
if SAVE_PLOTS:
	plt.savefig(timestamp+'_frequency_correction_fit.pdf', bbox_inches='tight')
# ----------------------------------------------------


# Adjust a sinusoidal function to each signal -------------------------------------
def sinusoidal_signal(t, Vp, f, phi):
	return Vp*np.sin(2*np.pi*f*t + phi)
def sinusoidal_signal_ff(t, p):
    return sinusoidal_signal(t, *p)
times_1 = np.linspace(0, SAMPLES_PER_BURST/fs1.val, SAMPLES_PER_BURST) # Array containing  sampling times
times_2 = np.linspace(0, SAMPLES_PER_BURST/fs2.val, SAMPLES_PER_BURST) # Array containing  sampling times
p0 = [GENERATOR_AMPLITUDE, GENERATOR_FREQUENCY, 0]
pfit, perr = fit_bootstrap(p0, times_1, samples_1, sinusoidal_signal_ff)
Vp1 = mg.magnitude(pfit[0], perr[0])
f1 = mg.magnitude(pfit[1], perr[1])
phi1 = mg.magnitude(pfit[2], perr[2])
pfit, perr = fit_bootstrap(p0, times_2, samples_2, sinusoidal_signal_ff)
Vp2 = mg.magnitude(pfit[0], perr[0])
f2 = mg.magnitude(pfit[1], perr[1])
phi2 = mg.magnitude(pfit[2], perr[2])

print('Voltmeter 1\n\tVp1 = ' + str(Vp1) + '\n\tf1 = ' + str(f1) + '\n\tphi1 = ' + str(phi1))
print('Voltmeter 2\n\tVp2 = ' + str(Vp2) + '\n\tf2 = ' + str(f2) + '\n\tphi2 = ' + str(phi2))
print(str(f1.val))
print('Differences:\n\tVp1 - Vp2 = ' + str(Vp1 - Vp2) + '\n\tf1 - f2 = ' + str(f1 - f2) + '\n\tphi1 - phi2 = ' + str(phi1 - phi2))

# PLOT DATA USING CORRECTION ON fs1 AND fs2 ----------------------------------------
FIG_WIDTH = 130e-3 # Figure width in meters (size of the chart+axis_ticks, not counting legend, title, etc.).
FIG_RATIO = [1, 0.4] # XY ratio aspect of figures.
f, (ax1) = plt.subplots(1, sharex=True, figsize=(FIG_WIDTH*FIG_RATIO[0]/25.4e-3, FIG_WIDTH*FIG_RATIO[1]/25.4e-3)) # Create the figure for plotting.
f.subplots_adjust(hspace=0.3) # Fine-tune figure; make subplots close to each other and hide x ticks for all but bottom plot.
ax1.plot(np.linspace(0, SAMPLES_PER_BURST/fs1.val, SAMPLES_PER_BURST), samples_1, marker='.', label='V1')
ax1.plot(np.linspace(0, SAMPLES_PER_BURST/fs2.val, SAMPLES_PER_BURST), samples_2, marker='.', label='V2')
ax1.plot(np.linspace(0, SAMPLES_PER_BURST/fs1.val, SAMPLES_PER_BURST), sinusoidal_signal(times_1, Vp1.val, f1.val, phi1.val), label='Fit for V1')
ax1.plot(np.linspace(0, SAMPLES_PER_BURST/fs2.val, SAMPLES_PER_BURST), sinusoidal_signal(times_2, Vp2.val, f2.val, phi2.val), label='Fit for V2')
ax1.set_xlabel('Time (s)')
ax1.set_ylabel('Voltage (V)')
ax1.grid(b=True, which='major', color='#aaaaaa', linestyle='-', linewidth=0.5) # Configure the grid in a more fachera way.
ax1.grid(b=True, which='minor', color='#dddddd', linestyle='-', linewidth=0.25) # Configure the grid in a more fachera way.
ax1.minorticks_on() # Enables minor ticks without text, only the ticks.
ax1.legend()
plt.show()
if SAVE_PLOTS:
	plt.savefig(timestamp+'_corrected_signals.pdf', bbox_inches='tight')
# ----------------------------------------------------


HP3458A_1.close()
HP3458A_2.close()
HP33120A.close()



When the line "HP33120A.write('OUTP:SYNC OFF')" is commented it works pretty fine, because the voltmeter triggers instantly after writing the command TARM SGL. But in this case the two voltmeters measure at different times. When that line is not commented I get a timeout in the line "HP3458A_1.write('TARM SGL')".

**Comment by MatthieuDartiailh** [comment]
Looking at the documentation of your instrument (following the link in your code https://xdevs.com/doc/HP_Agilent_Keysight/3458A/03458-90014.pdf) p254, it is mentioned (under Multiple arming) that `TARM SGL` blocks the GPIB line till all the triggers events have been received which means you cannot use it in the way you are trying to (and this is not a PyVISA limitation). For your application a trigger EXT may be more appropriate.

I am closing since there is nothing to be done on PyVISA end. Feel free to re-open should you have any related question ?

**Comment by SengerM** [comment]
Okay, thanks anyway. I have passed that paragraph away when configuring the voltmeter.
